proxy/stats_get_prediction.go

```diff
@@ -0,0 +1,345 @@
+			_, err := io.ReadAll(reader)
+			if err != nil {
+				fmt.Println(err)
+			}
+			delete(s.statsGetTasks, req)
+			return true
+		}
+		fmt.Println("Cache miss! " + req)
+		return false
+
+	}
+	return false
+}
+
+func (s *StatsGetPrediction) ParseLoginPrefix(loginRet []byte) {
+	s.loginPrefix = hex.EncodeToString(loginRet[60:79]) + hex.EncodeToString(loginRet[2:16])
+}
+
+func (s *StatsGetPrediction) BuildStatsReqBody(login string, req string) string {
+	/*
+
+		Get Stats Call Analysis
+		E.g.
+		data=9295xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx02a5302e302e350396a007ffffffff^@
+
+		1.
+		"data="
+		length=5
+
+		2. Header?
+		9295
+		length=2
+
+
+		3. Login 1?
+		index in login response=60
+		length=19
+
+		4. Login 2?
+		index in login response=2
+		length=14
+
+		5. Divider?
+		02a5
+		l=2
+
+		6. Version?
+		302e302e35 (0.0.5)
+		l=5
+
+		7. Divider2?
+		0396
+		l=2
+
+		8. Specific call
+		e.g. a007ffffffff , Confirm that this stays between users
+		l=6
+
+
+		9=End
+		\0
+		l=1
+	*/
+
+	body := "data=" +
+		"9295" + // Header
+		login +
+		"02a5" + // Divider
+		"302e302e35" + // 0.0.5
+		"0396" + // Divider 2
+		req +
+		"\x00" // End
+
+	return body
+}
+
+func (s *StatsGetPrediction) ProcessStatsQueue(queue chan *StatsGetTask) {
+	client := http.Client{
+		Transport: &http.Transport{
+			Proxy:                 http.ProxyFromEnvironment,
+			DialContext:           (&net.Dialer{Timeout: 30 * time.Second}).DialContext,
+			ResponseHeaderTimeout: 1 * time.Minute, // Some people have _really_ slow internet to Japan.
+			MaxIdleConns:          2,
+			MaxIdleConnsPerHost:   1,
+			MaxConnsPerHost:       2,
+			IdleConnTimeout:       90 * time.Second, // Drop idle connection after 90 seconds to balance between being nice to ASW and keeping things fast.
+			TLSHandshakeTimeout:   30 * time.Second,
+		},
+		Timeout: 3 * time.Minute, // 2x the slowest request I've seen.
+	}
+
+	for {
+		select {
+		case item := <-queue:
+			reqBytes := bytes.NewBuffer([]byte(item.request))
+			req, err := http.NewRequest("POST", s.GGStriveAPIURL+"statistics/get", reqBytes)
+			if err != nil {
+				fmt.Print("Req error: ")
+				fmt.Println(err)
+				item.response <- nil
+				continue
+			}
+			req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
+			req.Header.Set("Cache-Control", "no-cache")
+			req.Header.Set("Cookie", "theme=theme-dark")
+			req.Header.Set("User-Agent", "Steam")
+			req.Header.Set("Content-Length", strconv.Itoa(len(item.request)))
+
+			apiURL, err := url.Parse(s.GGStriveAPIURL) // TODO: Const this
+			if err != nil {
+				fmt.Print("Url error: ")
+				fmt.Println(err)
+				item.response <- nil
+				continue
+			}
+			apiURL.Path = req.URL.Path
+
+			req.URL = apiURL
+			req.Host = ""
+			req.RequestURI = ""
+			res, err := client.Do(req)
+
+			if err != nil {
+				fmt.Print("Res error: ")
+				fmt.Println(err)
+				item.response <- nil
+			} else {
+				item.response <- res
+			}
+		default:
```

Comment:
I think this causes a timing issue here. If the channel empties faster than it fills, it's possible that you'll return before you fetch everything. It's better to just block the for loop on the channel, and when the channel closes it will unblock.

Reply:
This shouldn't be an issue, since the queue is synchronously filled, then emptied after

Reply:
I still feel iffy about the pattern, but if it works, it works.